% This is samplepaper.tex, a sample chapter demonstrating the
% LLNCS macro package for Springer Computer Science proceedings;
% Version 2.20 of 2017/10/04
%
\documentclass[runningheads]{llncs}
%
\usepackage{minted}
\usepackage{pgf}
\usepackage{import}
\usepackage{layouts}
\usepackage{graphicx}
\usepackage{tikz}
\usetikzlibrary{shapes.geometric, arrows}
\usepackage{amsmath}
\usepackage{adjustbox}
\usepackage{placeins}
\usepackage{float}
\usepackage{xspace}
\usepackage{array}
\usepackage{tabularx,booktabs}\usepackage[hidelinks]{hyperref}
\usepackage{cleveref}
\usepackage{afterpage}
\usepackage{svg}
\usepackage{xcolor}
\usepackage{appendix}
\usepackage[nottoc]{tocbibind}
\usepackage[ngerman,english]{babel}
\newcolumntype{Y}{>{\centering\arraybackslash}X}


% Used for displaying a sample figure. If possible, figure files should
% be included in EPS format.
%
% If you use the hyperref package, please uncomment the following line
% to display URLs in blue roman font according to Springer's eBook style:
% \renewcommand\UrlFont{\color{blue}\rmfamily}

\renewcommand\#{\protect\scalebox{0.8}{\protect\raisebox{0.4ex}{\char"0023}}}
\definecolor{darkgrey}{RGB}{100,100,100} 
\newcommand{\sort}{SAMtools \texttt{sort}\xspace}
\newcommand{\points}{Data points: Median, Error bars: fastest and slowest of 3 runs. }
\newcommand{\parents}{Numbers in parentheses indicate the compression level. }
\newcommand{\threads}{One thread is single threaded computation; for a higher thread count, \sort uses the threads in addition to the main thread. }
\interfootnotelinepenalty=100000


\begin{document}

%
\title{Faster Sorting of Aligned DNA-Read Files}
%
%\titlerunning{Abbreviated paper title}
% If the paper title is too long for the running head, you can set
% an abbreviated paper title here
%
\author{Dominik Siebelt}
% Second Author\inst{2,3}\orcidID{1111-2222-3333-4444} \and
% Third Author\inst{3}\orcidID{2222--3333-4444-5555}}
%
\authorrunning{D.\@ Siebelt}
% First names are abbreviated in the running head.
% If there are more than two authors, 'et al.' is used.
%
\institute{Karlsruhe Institute of Technology}
% Springer Heidelberg, Tiergartenstr. 17, 69121 Heidelberg, Germany
% \email{lncs@springer.com}\\
% \url{http://www.springer.com/gp/computer-science/lncs} \and
% ABC Institute, Rupert-Karls-University Heidelberg, Heidelberg, Germany\\
% \email{\{abc,lncs\}@uni-heidelberg.de}}
%
\maketitle              % typeset the header of the contribution
%
\begin{abstract}
In the analysis of DNA sequencing data for finding disease causing mutations, to understand evolutionary relationships between species, and to find variants, DNA-Reads are compared to a reference genome, which is a representative example for a set of genes of a species. Sorting these aligned DNA-Reads by their position within the reference sequence is a crucial step in many of these downstream analyses. \sort, a widely used tool, performs external memory sorting of aligned DNA-Reads stored in the BAM format (Binary Alignment Map). This format allows for compressed storage of alignment data. \sort provides the most comprehensive set of features while exhibiting demonstrably faster execution times than its open source alternatives.  In this work, we analyze \sort for sorting BAM files and propose methods to reduce its runtime. We divide the analysis into three parts: management of temporary files, compression, and input/output (IO).
For the management of temporary files, we find that the maximum number of temporary files \sort can open concurrently is lower than the maximum number of open files permitted by the operating system. This results in an unnecessarily high number of merges of temporary files into larger temporary files, introducing overhead as \sort performs extra write and compression operations. To overcome this, we propose a dynamic limit for the number of temporary files, adapting to the operating system's soft limit for open files.
For compression, we test seven different libraries for compatible compression and a range of compression levels, identifying options that offer faster compression and result in a speedup of up to five times in single-threaded execution of \sort.
For IO, we demonstrate that a minimal level of compression avoids IO overhead, thereby reducing the runtime of \sort compared to uncompressed output. However, we also show that uncompressed output can be used in the pipelining of SAMtools commands to reduce the runtime of subsequent SAMtools commands.
Our proposed modifications to \sort and user behavior have the potential to achieve speedups of up to 6. This represents an important contribution to the field of bioinformatics, considering the widespread adoption of \sort evidenced by its over 5,000 citations and over 5.1 million downloads through Bioconda.
\end{abstract}

\begin{otherlanguage}{ngerman}
\begin{abstract}
    
\end{abstract}
\end{otherlanguage}
%
\setcounter{tocdepth}{2}
\tableofcontents
\newpage
%
\input{chapters/intro}
\input{chapters/prereqs}
\FloatBarrier
\newpage % Zwischen Chapter kann man sich ein Seitenumbruch g√∂nnen.
\input{chapters/related}
\newpage
\input{chapters/analysis}
\FloatBarrier
\newpage
\input{chapters/merging}
\newpage
\input{chapters/compression}
\FloatBarrier
\newpage
\input{chapters/io}
\newpage
\input{chapters/approaches}
\newpage
\input{chapters/conclusion}
\newpage

%
% ---- Bibliography ----
%
% BibTeX users should specify bibliography style 'splncs04'.
% References will then be sorted and formatted in the correct style.
%
\bibliographystyle{splncs04}
\bibliography{refs,references}
\input{chapters/appendix}
\end{document}
