\section{Introduction}

\subsection{Motivation}
Analysis of aligned DNA-Read data is a crucial part of modern bioinformatics with applications for scientific and medical purposes. DNA-Reads are short sequences of DNA bases encoding genetic information. They are aligned to a reference sequence to find variants, disease causing mutations and understand evolutionary relationships between species. In order to analyze specific parts of a genome, such as a single chromosome and to locate mutations or variations, DNA-Reads are sorted. This also improves the performance of many other downstream analysis tasks and algorithms, for example finding and removing duplicate reads. \\

To address the vast amounts of data generated by modern DNA sequencing machines, highly efficient tools and data formats are needed. Developed during the 1000 Genome Project~\cite{the_1000_genomes_project_consortium_1000_2012}, the SAM and BAM formats have found widespread use for storing alignment information to DNA sequences. Developed alongside these formats, SAMtools became a standard tool for manipulating SAM and BAM files. Beyond its extensive collection of commands for filtering, merging of aligned DNA-Read files and various other tasks, SAMtools provides the  \sort utility to reorder of the DNA sequences stored in DNA-Read files such as SAM and BAM files. \\

This study aims to optimize the performance of the \sort command for sorting DNA-Reads in BAM files. The primary objective is to reduce computational cost and processing time associated with this operation. BAM is a file format capable of storing alignment information to DNA sequences in a binary and compressed way. Given the substantial storage requirements of DNA data, the BAM file format incorporates compression to minimize storage costs, optimize storage capacity, and facilitate faster network transfer.\\

\sort can arrange DNA-Read files in various orders. The default order involves sorting aligned DNA-Reads by the reference sequence to which a DNA read is mapped, followed by the position of the mapping on this reference sequence. This enables fast random access to specific regions of interest via index files. \\

Index files are used in the handling of BAM files to facilitate random access despite compression. BAM files are compressed with a special compression method allowing to access the content of the file in blocks without the necessity of decompressing all preceding blocks of the file. The compression format internally uses the compression library zlib for GZIP for compression, which is one of the most popular compression methods. With zlib being open source and used extensively in various applications such as web servers and communication, other libraries have been implemented, offering GZIP compatible compression with higher throughput and higher compression efficiency. \\

In this thesis, I present three approaches to speed up the sorting of DNA-Reads stored in BAM files utilizing \sort. \\

\sort utilizes an external memory sorting algorithm. In situations with limited memory, temporary files are generated and subsequently merged, which is computationally demanding. In this thesis, I investigate the runtime performance implications of \sort writing and merging temporary files. I propose parameter settings and changes in \sort's internal limitation setting, reducing the amount of merges and thus lowering the runtime of \sort. \\

\sort dedicates a substantial amount of its runtime to compressing temporary and output files. To maintain the advantages of compression but reduce its impact on \sort's total runtime, I examine various zlib-compatible compression libraries and the effects of different compression levels. \\

Finally, I assess the impacts of IO operations and limitations of IO devices and propose recommendations to minimize or eliminate them.
