\section{Introduction}

\subsection{Motivation}
Analysis of aligned DNA-Read data is a crucial part of modern bioinformatics, with applications for scientific~\cite{derrien_fast_2012,kumar_snp_2016,logsdon_long-read_2020} and medical~\cite{lang_identification_2022,chen_personal_2012,yang_transmission_2017} purposes. DNA-Reads are short (typically 250â€“800 bases long) sequences of DNA bases encoding genetic information. They are aligned to a reference sequence, which serves as a standard DNA or RNA sequence for comparison within a species. The alignment can be used to find variants~\cite{tamura_virological_2023,pabinger_survey_2014,audano_characterizing_2019}, disease causing mutations~\cite{meyerson_advances_2010,he_comprehensive_2021,kim_cross-species_2020} and understand evolutionary relationships between species~\cite{mccormack_applications_2013,sexton_homology-based_2016,branstetter_enriching_2017}. In order to analyze specific parts of a genome, e.g., a single chromosome, DNA-Reads are sorted. This also improves the performance of many other downstream analysis tasks and algorithms~\cite{neph_bedops_2012,li_tabix_2011}, e.g., finding and removing duplicate reads~\cite{chen_gencore_2019,noauthor_samtools-markdup1_nodate}. 

To address the vast amounts of data (ranging from less than 10\,GB up to terabytes per sequencing run) generated by modern DNA sequencing machines, highly efficient tools and data formats are needed. Developed during the 1000 Genome Project~\cite{the_1000_genomes_project_consortium_1000_2012}, the SAM and BAM formats have found widespread use for storing alignment information to DNA sequences. Developed alongside these formats, SAMtools became a standard tool for manipulating SAM and BAM files. Beyond its collection of 38 commands for filtering, merging of aligned DNA-Read files and various other tasks, SAMtools provides the  \sort utility to reorder the DNA sequences stored in DNA-Read files such as SAM and BAM files. 

We aim to optimize the computational cost and processing time of the \sort command for sorting DNA-Reads in BAM files. BAM is a file format capable of storing alignment information to DNA sequences in a binary and compressed form. Given the substantial storage requirements (uncompressed BAM approximately 400\,GB per genome~\cite{noauthor_storage_2011}) of DNA data, the BAM file format incorporates compression to minimize storage costs, optimize storage capacity, and facilitate faster network transfer.

\sort sorts DNA-Read files by various sorting criteria, with the default order sorting aligned DNA-Reads by the ID of the reference sequence to which the DNA-Read is mapped, followed by the position of the mapping on this reference sequence. This enables fast random access to specific regions of interest via index files. 

BAM files are compressed with BGZF, a compression method allowing to access the content of the file in blocks without the necessity of decompressing all preceding blocks of the file. The compression format internally uses the popular zlib compression library for GZIP compression. Zlib is open source and used extensively in a wide array of applications (e.g., HTML-Compression). Thus, competing libraries have been implemented, offering GZIP compatible compression with higher throughput and smaller resulting files. 

In this thesis, we present three approaches to speed up the sorting of DNA-Reads stored in BAM files utilizing \sort: 

\sort utilizes an external memory sorting algorithm. In situations with limited memory, it utilizes temporary files and merges them subsequently. As writing temporary files requires additional computation steps, such as compression, writing temporary files is computationally demanding. Especially, merging temporary files into other temporary files invokes computational overhead, as \sort reads and decompresses already written temporary files and compresses and writes their content again. In this thesis, we investigate the runtime implications of \sort writing and merging temporary files. We propose parameter settings and changes in \sort's merging strategy, reducing the amount of merges and thus lowering the runtime of \sort. 

\sort dedicates a substantial amount of its runtime to compressing temporary and output files. To maintain the advantages of compression but reduce its impact on \sort's total runtime, we examine various GZIP-compatible compression libraries and the effects of different compression levels.

Finally, we assess the impacts of IO operations and limitations of IO devices and propose recommendations to minimize or eliminate IO blocklenecks. 
