\section{Related Work}

\subsection{Sorting Tools for aligned DNA-Read Files}
SAMtools is the most commonly used tool for manipulating SAM and BAM files, which store information on aligned DNA-Reads. With over 5.1~Million downloads (05.2024), it is the most downloaded package from the Bioconda~\cite{the_bioconda_team_bioconda_2018} channel, which is a common source for bioinformatic tools in the package and dependency manager Conda. However, there are some alternatives for sorting aligned DNA-Read files, such as Sambamba~\cite{tarasov_sambamba_2015}, Picard~\cite{Picard2019toolkit} and NovoSort~\cite{noauthor_novosort_nodate}. Picard, written in Java, provides a simpler interface for sorting SAM and BAM files, but its functions are limited compared to sort, and it does not prioritize efficiency. NovoSort is a commercial program and, therefore, I could not test it. Sambamba, written in the D programming language~\cite{alexandrescu_d_2010}, aims to provide a subset of SAMtools functionality, including sort, but with greater efficiency through higher parallelization. With SAMtools introducing parallelism in version 1.4 in 2017, the advantage Sambamba had over SAMtools diminished. For sorting, we report SAMtools 1.19 to be 3 times faster than Sambamba for small files (2.3\,GiB), both utilizing a total of 16 threads and 48\,GiB of memory. For larger files, the speedup of \sort compared to Sambamba \texttt{sort} increases to up to 5 at the largest input file we tested (213\,GiB). A comparison of the runtime of \sort and Sambamba \texttt{sort} on different input file sizes is shown in \Cref{fig:sambamba}.
\begin{figure}[t]
        \import{figures/}{sambamba.pgf}
    \caption{Runtime of Sambamba and SAMtools for sorting BAM files of different sizes. Both tools are default installations via Bioconda and utilize a total of 16 threads and 48\,GiB of memory. Data points represent median values across three replicate runs. Error bars depict the minimum and maximum values observed in these runs.}
    \label{fig:sambamba}
\end{figure}

\subsection{Alternative Compression Methods and File Formats}
Compression using BGZF, which is part of the specification of BAM files, utilizes GZIP. As a general purpose compression method, GZIP is not specialized in compressing DNA data. Hence, a variety of compression algorithms have been developed specifically for DNA data~\cite{hosseini_survey_2016}. These algorithms aim to minimize the resulting file size, compression speed, memory usage during compression or decompression, or the decompression speed, but typically focus primarily on minimizing the resulting file size. Compression algorithms are divided in \textit{reference-based} and \textit{reference-free} methods.

Reference-based algorithms utilize the reference sequence a DNA-Read is aligned to for compression. By only storing differences between each DNA-Read and the reference sequence, they archive better compression than reference-free methods. NGC~\cite{popitsch_ngc_2013} exemplifies this approach. It leverages reference sequences for compression and further enhances efficiency by splitting data types within a BAM file, like sequences, sequence names, and read qualities, into separate blocks for independent compression. Other algorithms and data formats like Goby~\cite{campagne_compression_2013}, DeeZ~\cite{hach_deez_2014}, and CRAM~\cite{fritz_efficient_2011} are also reference-based, but keep the advantage of enabling random access through index files. CRAM (Compressed Reference-oriented Alignment Map) is also built into HTSlib, the library SAMtools utilizes for file operations, and the user can choose to use it as output for \sort. NGC, Goby and CRAM also offer lossy compression. Lossy compression primarily targets meta information, including read qualities. Read qualities may not be crucial for every analysis and are independent of the reference sequence, making them less compressible. Thus, lossy compression of read qualities offers substantial storage savings.

Although reference compression methods for SAM files are still actively developed~\cite{banerjee_abridge_2022}, they lack interoperability, as the compression methods are not widely used. An exception to this is CRAM, which is supported by HTSlib. Although CRAM files offer several advantages over BAM files, including better compression, BAM files remain more popular due to their wider support among software tools.

Reference-free compression methods like BGZF, which per specification BAM files are compressed with, do not require the reference sequence for compression and decompression, making them more flexible, as e.g., no reference is required for sorting. Furthermore, reference-based compression methods often necessitate sorted DNA-Reads, making them inapplicable to inputs for \sort and outputs sorted after read names or other criteria. However, the only commonly used reference-free compression method is the BAM format, which uses binary encoding for the bases a DNA sequence consists of and BGZF compression. 

In summary, the BAM file format employing BGZF compression, which internally utilizes GZIP, is the most commonly used format for storing aligned DNA-Reads. Given its widespread use, optimizing the speed of generating BAM files using \sort is essential. However, increasing adoption of the CRAM format might lessen the future significance of BAM file generation.

